text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+  // we know our change output will be protected either by bootstrap outputs or by
+  // a protected input so we label it as a protected output
+  change_output = generate_protected_output(change_value)
+  outputs += [change_output]
+  return inputs, outputs
+}
+
+// TODO: needs handling of running out of protected outputs and keeping a minimal of `N`
+// protected outputs as defined in the configuration file
+```
+
+TODO: Check this is safe.
+
+#### Protection with wallet history
+
+A wallet can remember which outputs were spent. This way, if a spent output reappears in the wallet, a user is given a choice to either accept it or refresh it through a self-spend transaction.
+
+The downside of this approach is that a replay attacks are still possible in which case it would mean that a user would need to make a manual choice what to do with outputs that are unknown to the history of the wallet. Unknown outputs are also those created from the same seed on a different device and from child wallets since they each have their own history. The cross device scenario could be mitigated by export and import of wallet history and the child wallet outputs can be labeled and hence assumed as safe. It's up to the child wallet to protect itself from the attacks.
+
+### Receive-only wallets
+ 
+Any kind of automated receiving should default to 1-2 transactions and thus creating _unprotected_ outputs to avoid utxo spoofing attack which would reveal our inputs. Always performing 1-2 receive transactions can be achieved by setting the configuration `RECEIVE_PROTECTED_PROB = 0.0` which is explained in the next section.
+
+### Transaction building configuration
+
+```yaml
+// Defines a list of wallet configurations to define different behaviour that we do as a receiver of a
+// transactions. All wallet configurations inherit the values from the 'default' option that can be
+// overriden. A wallet should allow the user to generate a one-time SlatepackAddress based on a chosen
+// configuration. 
+TRANSACTION_BUILDING_CONFIGURATIONS:
+  default:
+    // Probability that we will protect a 'receive' transaction from replay attacks. If an anchor does not
+    // exist, the transaction will be protected by generating an anchor and an initial set of protected outputs.
+    // If such an anchor already exists, we contribute a protected output as an input which makes it a PayJoin
+    // transaction. By default, half of the receive transaction will be PayJoin transactions. This is to help mask
+    // the sender inputs. Additionally, PayJoin transactions are cheaper and more healthy for the network because
+    // they don't generate new outputs.
+    // If PayJoin transactions are not wanted due to privacy concerns, you can set this value to 0.0 in which case
+    // receive transactions will never contribute an input.
+    // Default: RECEIVE_PROTECTED_PROB = 0.5
+    RECEIVE_PROTECTED_PROB: 0.5

```

Comment:
I don't see why 0.5 is better than 1.0 as default.

Reply:
I was thinking about this and it's not yet clear to me what the consequences are. I personally prefer payjoins, but if you default to 1.0, then if you see a 2-2 transaction you can with high confidence assume that the inputs belong to different entities (not really sure what you can do with that data). If you have a 0.5, then it's much less likely that was the case because the probability of it being 2 inputs from the sender gets higher. I've not done analysis on this so I can't say whether it makes sense or not

Reply:
having to assume the inputs belong to different parties seems to be the worst case for a privacy attacker?!

Reply:
huh that's interesting. From the information theory perspective, you are giving him more information when you default to 1.0, but it seems like it's less useful information, so maybe you're right and higher values should be prefered :+1: (I'm still not sure 1.0 is optimal for privacy)